src/Features/ExternalAccess/HotReload/Api/HotReloadMSBuildWorkspace.cs

```diff
@@ -0,0 +1,237 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+extern alias BuildHost;
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+
+using MSB = Microsoft.Build;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.HotReload.Api;
+
+internal sealed partial class HotReloadMSBuildWorkspace : Workspace
+{
+    private readonly ILogger _logger;
+    private readonly MSBuildProjectLoader _loader;
+    private readonly ProjectFileInfoProvider _projectGraphFileInfoProvider;
+
+    public HotReloadMSBuildWorkspace(ILogger logger, Func<string, (ImmutableArray<MSB.Execution.ProjectInstance> instances, MSB.Evaluation.Project? project)> getBuildProjects)
```

Comment:
I wouldn't be entirely against if you wanted to make HotReloadMSBuildWorkspace abstract and convert this function to a virtual method; not sure if that might make it easier to change if you need to add new overloads or something. It's fine though.